crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,137 @@
+ */
+#include <stdio.h>
+/* The line above prevents "ISO C forbids an empty translation unit" warning */
+#ifndef FIPS_MODE
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "internal/bn_int.h"
+# include "../rand/rand_lcl.h"
+
+typedef struct drbg_ctx_st
+{
+    const unsigned char* x;
+    const unsigned char* h;
+    size_t rlen;
+} DRBG_CTX;
+
+static int app_data_index = 0;
+static size_t get_entropy(RAND_DRBG *drbg, unsigned char **pout,
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    DRBG_CTX *drbg_ctx = (DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)drbg_ctx->x;
+    return drbg_ctx->rlen;
+}
+
+static size_t get_nonce(RAND_DRBG *drbg, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
+{
+    DRBG_CTX *drbg_ctx = (DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)drbg_ctx->h;
+    return drbg_ctx->rlen;
+}
+
+static int bits2int(BIGNUM *out, int qlen,
+    const unsigned char *message, size_t message_len)
+{
+
+    if (BN_bin2bn(message, (int)message_len, out) == NULL)
+        return 0;
+    if ((int)message_len * 8 > qlen)
+        return BN_rshift(out, out, (int)message_len * 8 - qlen);
+    return 1;
+}
+static int int2octects(unsigned char *out, const BIGNUM  *num, int rlen)
+{
+    return BN_bn2binpad(num, out, rlen);
+}
+static int bits2octects(unsigned char *out, const BIGNUM  *range,
+    const unsigned char *message, size_t message_len, BN_CTX  *ctx)
+{
+    BIGNUM *num = BN_new();
+    if (
+        !bits2int(num, BN_num_bits(range), message, message_len) ||
+        !BN_mod(num, num, range, ctx) ||
+        !BN_bn2binpad(num, out, BN_num_bytes(range)))
+    {
+        BN_free(num);
+        return 0;
+    }
+    BN_clear_free(num);
+    return 1;
+}
+int bn_generate_dsa_deterministic_nonce(BIGNUM *out, const BIGNUM *range,
+    const BIGNUM *priv, const unsigned char *message,
+    size_t message_len, int hash_type, BN_CTX *ctx)
+{
+    int ret = 0;
+
+    RAND_DRBG *drbg = NULL;
+    DRBG_CTX drbg_ctx;
+    unsigned int drbg_flags = RAND_DRBG_FLAG_HMAC;
+    int rlen = 0, qlen = 0;
+    unsigned char *x = NULL, *h = NULL, *T = NULL;
+
+    app_data_index = RAND_DRBG_get_ex_new_index(0L, NULL, NULL, NULL, NULL);
+
+    if ((qlen = BN_num_bits(range)) < 0
+        || (rlen = BN_num_bytes(range)) < 0)
+        goto cleanup;
+
+    if ((drbg = RAND_DRBG_new(hash_type, drbg_flags, NULL)) == NULL
+        || !RAND_DRBG_set_callbacks(drbg, get_entropy, NULL, get_nonce, NULL))
+        goto cleanup;
+
+    if ((x = (unsigned char *)OPENSSL_malloc(rlen)) == NULL
+        || (h = (unsigned char *)OPENSSL_malloc(rlen)) == NULL
+        || (T = (unsigned char *)OPENSSL_malloc(rlen)) == NULL
+        || (memset(&drbg_ctx, 0, sizeof(drbg_ctx)) == NULL))
```

Comment:
The call to memset(3) cannot fail, there is no need to check its result.

Reply:
Sollved